### DISCUSSION TITLE
Rendering approaches?

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#71] [Ideas]
This is a biggie. A major undertaking.

We can take a number of approaches:

1. Ignore rendering for now.
2. Use USD export to transfer scenes to Blender et al.
3. Link to an existing open source renderer (C/C++).
4. Write our own renderer (though I suspect the intersection of Lisp devs and rendering devs is close to the empty set).

Thoughts?

**Comment by shirok** [comment]
For the batch rendering, exporting to (and possibly running as a subprocess) external renderer would be the practical approach.  For the real-time preview, it would be fun to write your own---starting from very primitive ones and adding fancier features gradually---for it's something you'll always use while developing.  Just my $.02

**Reply by kaveh808** [reply]
It would be fun if we can think of some Lispy or otherwise interesting thing to do. That's a touchstone of mine for the project: how can we be different and better?

I wonder if real-time rendering these days is just packing data off to the GPU, or if there is room for design.

**Comment by aykaramba** [comment]
Here is someone that wrote a small raytracer in Lisp as a toy project:  https://www.youtube.com/watch?v=N1oMRw04W3E

Here is his github: https://github.com/afwlehmann/clrt-tutorial

**Reply by kaveh808** [reply]
Nice motivation, but probably not applicable for our needs.

**Comment by gbanel** [comment]
Maybe something like shadertoy could be a good option for the project. A shader based render would give a more interesting look (than usual polys) and linking it to lisp could help create infinite variations. A few ideas that would work:
https://www.shadertoy.com/view/4dfGzs
https://www.shadertoy.com/view/ssjyWc
https://www.shadertoy.com/view/wsGfD3
https://www.shadertoy.com/view/MdcfWH
https://www.shadertoy.com/view/lsjGRV
https://www.shadertoy.com/view/4dcGW2
https://www.shadertoy.com/view/XsBXWt
https://www.shadertoy.com/view/XslGRr
https://www.shadertoy.com/view/MdX3Rr
https://www.shadertoy.com/view/7lffDr

**Reply by kaveh808** [reply]
Interesting stuff. Do people write these shaders by hand or are there authoring tools?

And is there a way of rendering geometry with this, or is this only for pixels?

**Reply by gbanel** [reply]
In shadertoy, it's only pixels shaders on a flat image, not much geometry, apart from implicit geometry.
It's done manually, no tools to generate it, it's a game, and some people create amazing stuff:
https://www.shadertoy.com/view/lslfRN

**Comment by kaveh808** [comment]
It would be interesting to see if we can wrap this renderer in FFI and call it from our system:

https://luxcorerender.org

Anyone care to investigate?

**Reply by ghost** [reply]
I was investigating this yesterday.  One issue is that they went to a C++/Python API at 2.0.  To use CFFI you would have to flatten the API to C.  They have a scene description language (SDL), which could be generated by kons-9 as an interop.

**Comment by jolby** [comment]
If you want to experiment with a real-time physical based rendering engine, Pavel Korolev (borodust) has already created CL bindings to google's Filament renderer:
https://github.com/borodust/aw-filament

You can see his usage of the wrapped renderer in his alien-works game engine project:
https://github.com/borodust/alien-works
An example game created with the alien-works engine:
https://github.com/borodust/notalone-thriced

Lots of good graphics stuff in his repos.

**Reply by ghost** [reply]
borodust apparently used CLAW to wrap Filament.  CLAW has gone from "C++ support pending" to "Still BETA quality. API is subject to change".  Maybe CLAW could be used for wrapping LuxCore.

**Reply by kaveh808** [reply]
This looks interesting. Let's explore.

**Reply by kaveh808** [reply]
Have you found any examples of calling Filament from CL? I couldn't find any.

**Reply by jolby** [reply]
@kaveh808 If you look in the graphics subfolder of borodust's alien-works project you will see examples of how he calls the filament api using the raw wrapped bindings. Check out the filament subfolder:
https://github.com/borodust/alien-works/tree/master/src/graphics/filament

And also the file engine.lisp where he defines a renderer interface based on those lower level bindings:
https://github.com/borodust/alien-works/blob/master/src/graphics/engine.lisp

**Reply by kaveh808** [reply]
Thanks for finding these. I contacted the developer asking if there were any examples of CL-filament use, and he said not currently, but he will have some at some point.

If anyone has the knowledge and is interested in digging more into this, please let me know and I'll make an issue.

**Comment by JMC-design** [comment]
depending on how much interactivity you need, can just call it using python or the shell.
Another case for abstraction.

**Reply by kaveh808** [reply]
I did look at https://luxcorerender.org

They have sample scenes on their site, but the geo (.ply files) is in binary format. If we can find ASCII samples, it may be an interesting test to export scene data and trigger rendering via the shell.

**Reply by kaveh808** [reply]
I did manage to export ASCII .ply data from Blender. Might write a LuxCore exporter from kons-9 at some point.

**Reply by JMC-design** [reply]
I was thinking of making a general ply reader/writer. It is such a simple
yet extensible format that coming up with a good general solution is a bit
hard as there seems to be very little info on what standard keywords are.
Just how many slots does a vertex have, do they add in colours,normals, and
materials all in one?
So the general case might take time for an eloquent solution, if all you
need to do is output it becomes easy...if you know what fields the receiver
expects. It's just a list of objects, their number and description,
followed by the data in same order of definition. The binary format is the
data as layed out in memory.
this would lead to an easy export if you did something like array of
structs for rendering. though probably wouldn't be so hard the other way
either.
